jina/types/arrays/neural_ops.py

```diff
@@ -248,3 +247,48 @@ def visualize(
             plt.savefig(output, bbox_inches='tight', pad_inches=0.1)
         else:
             plt.show()
+
+    @property
+    def embeddings(self) -> np.ndarray:
+        """Return a `np.ndarray` stacking all the `embedding` attributes as rows.
+
+        Warning: This operation assumes all embeddings have the same shape.
+
+        Warning: This operation currently does not support sparse arrays.
+
+        :return: embeddings stacked per row as `np.ndarray`.
+        """
+
+        x_mat = b''.join(d.proto.embedding.dense.buffer for d in self)
+
+        return np.frombuffer(x_mat, dtype=self[0].proto.embedding.dense.dtype).reshape(
+            (len(self), self[0].proto.embedding.dense.shape[0])
+        )
+
+    def _get_embeddings(self, indices: Optional[slice] = None) -> np.ndarray:
+        """Return a `np.ndarray` stacking  the `embedding` attributes as rows.
+        If indices is passed the embeddings from the indices are retrieved, otherwise
+        all indices are retrieved.
+
+        Example: `self._get_embeddings(10:20)` will return 10 embeddings from positions 10 to 20
+                  in the `DocumentArray` or `DocumentArrayMemmap`
+
+        Warning: This operation assumes all embeddings have the same shape.
+
+        Warning: This operation currently does not support sparse arrays.
+
+        :param indices: slice of data from where to retrieve embeddings.
+        :return: embeddings stacked per row as `np.ndarray`.
+        """
+        if indices is None:
+            indices = slice(0, len(self))
+
+        x_mat = bytearray()
+        len_slice = 0
+        for d in self[indices]:
+            x_mat += d.proto.embedding.dense.buffer
```

Comment:
I would prefer to use `assert` to guarantee all of embedding is dense. 

```suggestion
            assert d.proto.embedding.dense, 'does not support sparse arrays'
            x_mat += d.proto.embedding.dense.buffer
```

Reply:
That would kill performance. Operations at this level are not usually checked.

Reply:
I agree that the checking procedure should happen at the DocumentArray level.